First steps to provides.py for elastic search

sebastienpattyn93 opened the pull request:
## overview

- Updated Readme for provides.py 
- added provides.py

#### provides.py
 - host, clustername and port are set
- Elasticsearch charm will have to provide these

##### Elastic Search
- currently working on a reactive layered elasticsearch(version 5.1) charm where this layer is being used to make relationships with the metricbeat charm (also in development)

chuckbutler commented:
I haven't tested this code in anger yet, but preliminary review: This is good stuff! 

I'll get this PR added to my watch list and get you a proper review in the next day or so. Thanks for the submission @sebastienpattyn93

jamesbeedy commented on provides.py:
```diff
@@ -0,0 +1,44 @@
+        conv = self.conversation()
+        conv.set_remote(data={
+            'port': port,
+            'cluster_name': cluster_name,
```
@sebastienpattyn93 nice!

sebastienpattyn93 commented:
@chuckbutler thanks for the feedback. Hadn't thought of that yet. I changed the provides.py events that they match now. Still thinking if it's necessary to make a difference here between ready and connected state and that it might be an option to combine {joined,changed} into one event that sets the connected state

chuckbutler commented:
That sounds good. 

I have one pedantic thing, and feel free to fight me on this one because i can be easily swayed here...

we have a set of states already documented in the README that give a clear outline of what each represents

```
{relation_name}.connected - Denotes that the client has connected to the Elasticsearch node(s), but has not yet received the data to configure the connection.

{relation_name}.available - Denotes that the client has connected and received all the information from the provider to make the connection.

{relation_name}.departed - Denotes that the unit has departed from the Elasticsearch relationship, and should be removed from any configuration files, etc.
```

Should we maybe not use `.ready` and use `.available` since it matches on the requires side? or is the `.ready` signifying something slightly different and thus requires a different state name?

sebastienpattyn93 commented:
It does make more sense when I look at it that way. I added it in earlier stages since I was developing it next to an elasticsearch layered charm. But I have not yet figured out where an extra stage would be necessary. I made it to one event as well so only connected and broken are necessary at the moment. I could then add the `.available` if I want to check something on the other side, but still looking at what would be the best since this interface might be used by beats and or Kibana

chuckbutler commented:
I would think beats/kibana only needs the 'requires.py'. As 'provides.py' is intended for the unit providing elasticsearch connectivity details.

sebastienpattyn93 commented:
Then only the `.connected` and `.broken` states should be needed to make it work on the `provides.py` side

chuckbutler commented:
I am +1 to these changes. Thanks for the submission @sebastienpattyn93
